### DISCUSSION TITLE
How augmentations work

### DISCUSSION THREAD

**Discussion by AwayFromKeyb0ard** [ultralytics/ultralytics#10018] [Q&A]
Hey there,

i want to ask about augmentations. is there any difference if i use augment.py file to add my augmentations or use hyperparameters in default.yaml file? 

When i use albumentations to do augmentation in my code during training i see that if i multiply batch size with num of batches gaves me the number of original images in every epoch. The augmented images are included in every training? Because i cannot see how many are created. The only thing i did is to add more augmentations in augment.py in class albumentations and i can see them when i do 
yolo detect train data=coco128.yaml model=yolov8n.yaml pretrained=yolov8n.pt epochs=100 imgsz=640.

**Comment by pderrenger** [comment]
Hello there! 😊

You can indeed use either the augmentations defined in `augment.py` or specify hyperparameters directly in the `default.yaml` for customization. However, they serve slightly different purposes:

1. **Modifying `augment.py`**: Allows you more flexibility and the ability to use the Albumentations library for complex augmentation pipelines directly in your code. This is especially useful if you need augmentations that aren't covered by the basic hyperparameters in the YAML file.

2. **Using hyperparameters in `default.yaml`**: This method lets you conveniently adjust the implementation and intensity of built-in augmentations, like rotation or translation, without directly modifying the code.

Regarding how augmentations are applied: When training, the augmentations are applied on-the-fly. This means they're not pre-generated or stored but instead created in real-time during each epoch. So, if you've added augmentations via `augment.py`, they are indeed being applied, although the exact count of augmented images is not explicitly reported.

You can observe the effects of the augmentations during training by checking the logged outputs or saved training samples, which can give you insight into the variety of transformations being applied.

Hope this clarifies your query! Let us know if there's anything else you're curious about.

**Comment by AwayFromKeyb0ard** [comment]
Thank you for your response @pderrenger ,

So, augmentations make "virtual" new photos to train the model alongside with the original photos? For example, if X is the number of original photos in my dataset and augmentations create Y images, the model totally is trained on X+Y images?

**Reply by pderrenger** [reply]
Hey there! 😊

Absolutely correct! When you augment your dataset with additional transformations, it's like you're adding "virtual" images to your training set. So, if your original dataset contains X images and augmentations create Y additional transformed images, your model effectively sees and learns from a total of X+Y images during training. These augmentations help in making your model more robust by introducing a variety of conditions it might encounter in real-world scenarios. 

Hope this clarifies your query! Happy training! 🚀

**Reply by Ishfaz** [reply]
Hello, 
If  I want to add more augmentations, such as blur, noise, contrast, and brightness, which are not built into the augment.py file? Should these be added directly in the augment.py file, or should they be incorporated during the Via callback function?

**Reply by glenn-jocher** [reply]
For custom augmentations like blur/noise/contrast, we recommend modifying the Albumentations transforms list directly in the `Albumentations` class within augment.py ([source](https://docs.ultralytics.com/reference/data/augment/#albumentations)). You can add new Albumentations transforms to the `T` list while maintaining the existing YOLO bbox handling. The Ultralytics team generally advises against callback modifications for core augmentation functionality. 

For implementation specifics, refer to the Albumentations documentation of supported transforms: https://albumentations.ai/docs/

**Reply by Ali-Raza12** [reply]
@glenn-jocher correct me if I am wrong. Isn't that augmentations are dynamically applied during training iteration and this process is random so the model sees different version of base image in every iteration rather than augmentation being applied on X base images to make a total dataset of X+Y Images before training starts?

**Reply by glenn-jocher** [reply]
You're absolutely correct! Augmentations in Ultralytics YOLO are applied dynamically during training iterations, generating new randomized variations of each image in real-time rather than creating a static augmented dataset. This means the model sees fresh variations of every image in each epoch, maximizing data diversity without permanently expanding the dataset size. For implementation details, see the Albumentations class in our [augment.py source code](https://docs.ultralytics.com/reference/data/augment/#ultralytics.data.augment.Albumentations).

**Reply by marcofarrugia** [reply]
@glenn-jocher, does the framework guarantee the original (unmodified) images will be used in at least one of the epochs?
